### PR TITLE
[Event Hubs] Pluggable Checkpoint Store

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The focus of these changes is to build on the `CheckpointStore` contract, exposing a BlobCheckpointStore and PluggableCheckpointStoreEventProcessor<T> that can be used to create custom processors without the need to implement storage functionality.

# References and Related

- [[Event Hubs] Checkpoint Store Abstraction (#27386)](https://github.com/Azure/azure-sdk-for-net/pull/27386)
- [Event Hubs: Checkpoint Store Proposal](https://gist.github.com/jsquire/bfbbe8b0d9f27d3d5b9c63ceb9390ce8)

### CODE REVIEW

**Comment by jsquire** [comment]
azp run net - eventhub - tests

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PluggableCheckpointStoreEventProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        ///   which is needed.  In this case, the name can be added manually by adding ";EntityPath=[[ EVENT HUB NAME ]]" to the end of the
+        ///   connection string.  For example, ";EntityPath=telemetry-hub".
+        ///
+        ///   If you have defined a shared access policy directly on the Event Hub itself, then copying the connection string from that
+        ///   Event Hub will result in a connection string that contains the name.
+        /// </remarks>
+        ///
+        /// <exception cref="ArgumentOutOfRangeException">Occurs when the requested <paramref name="eventBatchMaximumCount"/> is less than 1.</exception>
+        ///
+        /// <seealso href="https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-get-connection-string">How to get an Event Hubs connection string</seealso>
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PluggableCheckpointStoreEventProcessor.cs:55]:
```suggestion
        /// <seealso href="https://docs.microsoft.com/azure/event-hubs/event-hubs-get-connection-string">How to get an Event Hubs connection string</seealso>
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PluggableCheckpointStoreEventProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        ///
+        /// <remarks>
+        ///   If the connection string is copied from the Event Hub itself, it will contain the name of the desired Event Hub,
+        ///   and can be used directly without passing the <paramref name="eventHubName" />.  The name of the Event Hub should be
+        ///   passed only once, either as part of the connection string or separately.
+        /// </remarks>
+        ///
+        /// <exception cref="ArgumentOutOfRangeException">Occurs when the requested <paramref name="eventBatchMaximumCount"/> is less than 1.</exception>
+        ///
+        /// <seealso href="https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-get-connection-string">How to get an Event Hubs connection string</seealso>
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PluggableCheckpointStoreEventProcessor.cs:86]:
```suggestion
        /// <seealso href="https://docs.microsoft.com/azure/event-hubs/event-hubs-get-connection-string">How to get an Event Hubs connection string</seealso>
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PluggableCheckpointStoreEventProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        ///   be used to initialize the partition for processing.
+        ///
+        ///   In the event that a custom starting point is desired for a single partition, or each partition should start at a unique place,
+        ///   it is recommended that this method express that intent by returning checkpoints for those partitions with the desired custom
+        ///   starting location set.
+        /// </remarks>
+        ///
+        protected override async Task<EventProcessorCheckpoint> GetCheckpointAsync(string partitionId,
+                                                                                   CancellationToken cancellationToken) =>
+            await _checkpointStore.GetCheckpointAsync(FullyQualifiedNamespace, EventHubName, ConsumerGroup, partitionId, cancellationToken).ConfigureAwait(false);
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PluggableCheckpointStoreEventProcessor.cs:206]: Is it important that we `await` the call here or could we avoid the extra state machine and return the Task directly?

**Review by christothes — APPROVED** [review]

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PluggableCheckpointStoreEventProcessor.cs (hunk 3 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/PluggableCheckpointStoreEventProcessor.cs:206]: I don't think that it's important; but doesn't our general guidance call for awaiting everywhere?   Personally, I don't think that seeing `GetCheckpointAsync` in the stacktrace is helpful.

**Pull request merged**